File: C++/Backtracking/nqueen.cpp
#include <iostream>
using namespace std;

bool canPlace(int board[][20], int n, int x, int y)
{

    
   /* Checking if there is a queen in the same column. */
    for (int k = 0; k < x; k++)
    {
        if (board[k][y] == 1)
        {
            return false;
        }
    }
    
   /* Checking if there is a queen in the left diagonal. */
    int i = x;
    int j = y;
    while (i >= 0 and j >= 0)
    {
        if (board[i][j] == 1)
        {
            return false;
        }
        i--;
        j--;
    }

    
    /* Checking if there is a queen in the right diagonal. */
    i = x;
    j = y;
    while (i >= 0 and j < n)
    {
        if (board[i][j] == 1)
        {
            return false;
        }
        i--;
        j++;
    }
    return true;
}

void printBoard(int n, int board[][20])
{

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cout << board[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl;
}

/**
 * We are trying to place a queen in every row, and for every queen we place, we are trying to place
 * the next queen in the next row. 
 * 
 * If we are able to place all the queens, we have found a solution. 
 * 
 * If we are not able to place all the queens, we backtrack and try to place the queen in the next
 * column. 
 * 
 * If we are not able to place the queen in any column, we backtrack and try to place the queen in the
 * next column in the previous row. 
 * 
 * If we are not able to place the queen in any column in the previous row, we backtrack and try to
 * place the queen in the next column in the previous row. 
 * 
 * And so on.
 * 
 * Args:
 *   n (int): The size of the board.
 *   board (int): The chess board.
 *   i (int): row number
 * 
 * Returns:
 *   a boolean value.
 */
bool solveNQueen(int n, int board[][20], int i)
{
    // base case
    if (i == n)
    {
        // Print the board
        printBoard(n, board);
        return true;
    }

    // rec case
    // try to place a queen in every row
    for (int j = 0; j < n; j++)
    {
        // whether the current i,j is safe or not
        if (canPlace(board, n, i, j))
        {
            board[i][j] = 1;
            bool success = solveNQueen(n, board, i + 1);
            if (success)
            {
                return true;
            }
            // backtrack
            board[i][j] = 0;
        }
    }
    return false;
}
/**
 * If the current queen can be placed safely in current column, mark this column as part of the
 * solution and recursively check if placing a queen in the next column leads to a solution
 * 
 * Returns:
 *   The number of solutions to the N-Queen problem.
 */
int main()
{
    int board[20][20] = {0};
    int n;
    cout<<"Number of Queens you want to place : ";
    cin >> n;
    solveNQueen(n, board, 0);
    return 0;
}

File: Python/Divide-and-Conquer/N_Queen.py
# Python3 program to solve N Queen
# Problem using backtracking

global N
N = 4


def printSolution(board):
	for i in range(N):
		for j in range(N):
			if board[i][j] == 1:
				print("Q",end=" ")
			else:
				print(".",end=" ")
		print()


# A utility function to check if a queen can
# be placed on board[row][col]. Note that this
# function is called when "col" queens are
# already placed in columns from 0 to col -1.
# So we need to check only left side for
# attacking queens
def isSafe(board, row, col):

	# Check this row on left side
	for i in range(col):
		if board[row][i] == 1:
			return False

	# Check upper diagonal on left side
	for i, j in zip(range(row, -1, -1),
					range(col, -1, -1)):
		if board[i][j] == 1:
			return False

	# Check lower diagonal on left side
	for i, j in zip(range(row, N, 1),
					range(col, -1, -1)):
		if board[i][j] == 1:
			return False

	return True


def solveNQUtil(board, col):

	# Base case: If all queens are placed
	# then return true
	if col >= N:
		return True

	# Consider this column and try placing
	# this queen in all rows one by one
	for i in range(N):

		if isSafe(board, i, col):

			# Place this queen in board[i][col]
			board[i][col] = 1

			# Recur to place rest of the queens
			if solveNQUtil(board, col + 1) == True:
				return True

			# If placing queen in board[i][col
			# doesn't lead to a solution, then
			# queen from board[i][col]
			board[i][col] = 0

	# If the queen can not be placed in any row in
	# this column col then return false
	return False


# This function solves the N Queen problem using
# Backtracking. It mainly uses solveNQUtil() to
# solve the problem. It returns false if queens
# cannot be placed, otherwise return true and
# placement of queens in the form of 1s.
# note that there may be more than one
# solutions, this function prints one of the
# feasible solutions.
def solveNQ():
	board = [[0, 0, 0, 0],
			[0, 0, 0, 0],
			[0, 0, 0, 0],
			[0, 0, 0, 0]]

	if solveNQUtil(board, 0) == False:
		print("Solution does not exist")
		return False

	printSolution(board)
	return True


# Driver Code
if __name__ == '__main__':
	solveNQ()


